docs: document more ways to secure the data api (#27756)
supabase/supabase: apps/docs/content/guides/database/hardening-data-api.mdx

---
title: 'Hardening the Data API'
---

> Your database's automatically generated Data API exposes the `public` schema by default. Since the `public` schema is often used by other tools as a default space, you might want to lock down this schema. This helps prevent accidental exposure of data that's automatically added to `public`.

Reviewer: ```suggestion
Your database's automatically generated Data API exposes the `public` schema by default. If your `public` schema is used by other tools as a default space, you might want to lock down this schema. This helps prevent accidental exposure of data that's automatically added to `public`.
```

Any data, views, or functions that should be exposed need to be deliberately put within your custom schema (which we will call `api`), rather than ending up there by default.

### Step 1: Remove `public` from exposed schemas

1.  Go to [**API Settings**](/dashboard/project/_/settings/api) in the Supabase Dashboard.
1.  Under **Data API Settings**, remove `public` from **Exposed schemas**. Also remove `public` from **Extra search path**.
1.  Click **Save**.
1.  Go to [**Database Extensions**](/dashboard/project/_/database/extensions) and disable the `pg_graphql` extension.

### Step 2: Create an `api` schema and expose it

1.  Connect to your database. You can use `psql`, the [Supabase SQL Editor](/dashboard/project/_/sql), or the Postgres client of your choice.

1.  Create a new schema named `api`:

    ```sql
    create schema if not exists api;
    ```

1.  Grant the `anon` and `authenticated` roles usage on this schema. Also grant any other permissions they will need, such as ability to read and edit tables:

    ```sql
    grant usage on schema api to anon, authenticated;

>     grant select, insert, update, delete on all tables in schema api to anon, authenticated;

Reviewer 2: This would be an insecure setup. `anon` should not be granted any privilege by default. `authenticated` should also be granted privileges explicitly for each table and operation.

As it is users will just copy paste this and end up with a similar setup as Supabase defaults.

For now I suggest:

```suggestion
  grant select on table <your_table> to anon; -- only read access for anon users  
  grant select, insert, update, delete on table <your_table> to authenticated; -- read and write access for authenticated users  
```
